daffodil-core/src/main/scala/org/apache/daffodil/compiler/Compiler.scala

```diff
@@ -352,15 +351,15 @@ class Compiler private (var validateDFDLSchemas: Boolean,
     val diags = pf.getDiagnostics // might be warnings even if not isError
     if (err) {
       Assert.invariant(diags.nonEmpty)
-      log(LogLevel.Compile, "Compilation (ProcessorFactory) produced %d errors/warnings.", diags.length)
+      Logger.log.debug(s"Compilation (ProcessorFactory) produced ${diags.length} errors/warnings.")
```

Comment:
This style fundamentally bothers me. It is incompatible with Internationalization of the logging. 
However, unlike error messages like our SDE/SDW/PE/UE/VE errors, logging quite possibly should be English only, so this is ok, 

I just like the idea of having only one idiom for messaging - whether it is logging, issuing a SDE, or actually printing something out. 

I always prefer this style:
```
loggingOrOtherMesageCreator("Message in English text with %s placeholders", arg0, arg1, arg2...argK)
```
as the style because it enables internationalization by using the Message string as a lookup key into a database by desired locale, and the object retrieved can scrutinize arguments to choose correct text.

Reply:
Unfortunately, the Scala API logger doesn't support this args:

https://logging.apache.org/log4j/scala/api/2.12/org/apache/logging/log4j/scala/Logger.html

I'm not sure if this is a limitation of their macro implementation, or if it's just missinng, but that's not really an option.

To support that, we would probably need to add our own wrapers, or do something like `"message %s".format(args)`, which I personally find kind of ugly, but I'm not that strongly against it.

Though, there is a benefit to use Scala's string interpolation over `format`, which is that it's impossible to create a format string that doesn't match the parameters. For example, if you do something like this
```scala
"Hello: %s %s".format(oneParam)
```
Then you end up getting an `MissingFormatArgumentException` at runtime. Using string interpolation, that's not possible. So there is a class of errors that simply go away.

Also, regarding internationalization, can you just internationalize the entire string? It's maybe bit more verbose, but there isn't that much of a difference between `Bonjour: %s` and `Bonjour: %{oneParam}`, aside from needing to be careful to not translate variable names.

I also wonder if string interpolation would allow for more complex internationalization logic? For example, you could maybe internationalize ordinals with something like:

```scala
s"This is the ${ordinalString("english", num)} thing"
s"C'est la ${ordinalString("french", num)} chose"
```
I don't really know much about internationalization, or how things like ordinals are usually handled, so maybe this makes no sense. But maybe does shows some flexibility over simple format strings.

Reply:
I take your point that string interpolation is better because it rules out a class of errors at compile time, which I like. 

Suppose the macro expanded so that it passes both the raw string (no interpolated substitutions) and the evaluated string (with substitutions) to the underlying messaging layer.

Then I think the underlying layer can handle the internationalization just fine. By comparing the raw and substituted strings it can extract the args to the string interpolation, for use with an internationalized version of the text.

So a more clever macro is the only thing needed. No source code changes. That gets rid of my concerns about this.

Reply:
Thinking some one must have already done this, I did find this reddit thread about someone doing an internationalization string interpolator:

https://www.reddit.com/r/scala/comments/31iey4/scala_string_interpolation_why_the_s/

It doesn't use macros, and has some overhead to do things like rebuild the raw string, but I think the idea is sound. A macro implementation would probably be necessary once we wanted to support internationalization.

Another befit of using a string interpolator is that if we had something like `i"Internationalizable message: $foo"`, then it would fairly east to find all the `i"..."` strings that need to be translated to build the database.

Reply:
Good. So I think we need to do nothing right now. I just wanted to not go down a path that would leave us with no options.

Seems like we should move in this direction for other messages like for SDE and such also, so that this interpolator/macro technique would apply there as well. In the past we have had some bugs due to missing args to match a "%s" in the format string, or vice versa where there were args, but no corresponding "%s" in the format string. 

Properly using string interpolation with macros would fix this, yet still be amenable to internationalization. Something to consider.

Reply:
Good point. I'll create some tickets to update SDE's and to consider some internalization string interpolator.